### ISSUE TITLE
Newbie Q: is it possible to train n (look ahead) tokens at a time?

### ISSUE THREAD

**Issue by mw66** [karpathy/nanoGPT#378]
Hi,

Right now, at each training step, the `y` is ahead of `x` by one token, I'm just wondering if we can train 2 or even more future tokes in `y` at each step?

Will this affect training speed and quality? and what are the necessary change needed for such to happen?

thanks.

**Comment by VatsaDev** [comment]
The generation code works like this:

```python
@torch.no_grad()
    def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None):
        """
        Take a conditioning sequence of indices idx (LongTensor of shape (b,t)) and complete
        the sequence max_new_tokens times, feeding the predictions back into the model each time.
        Most likely you'll want to make sure to be in model.eval() mode of operation for this.
        """
        for _ in range(max_new_tokens):
            # if the sequence context is growing too long we must crop it at block_size
            idx_cond = idx if idx.size(1) <= self.config.block_size else idx[:, -self.config.block_size:]
            # forward the model to get the logits for the index in the sequence
            logits, _ = self(idx_cond)
            # pluck the logits at the final step and scale by desired temperature
            logits = logits[:, -1, :] / temperature
            # optionally crop the logits to only the top k options
            if top_k is not None:
                v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
                logits[logits < v[:, [-1]]] = -float('Inf')
            # apply softmax to convert logits to (normalized) probabilities
            probs = F.softmax(logits, dim=-1)
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1)
            # append sampled index to the running sequence and continue
            idx = torch.cat((idx, idx_next), dim=1)

        return idx
```

Basically, we loop through the context `max_new_tokens` times, 
 - checking if were hitting the context limit, if so, crop it
 - ask the model for the next token, checking fo top_k crop
 - sampling the token
 - concatenating it to the current context

We might be able to get multiple tokens at a time with `idx_next = torch.multinomial(probs, num_samples=1)` being set to 2 or 5, etc, gotta test that.

but still, it won't be the same thing you're suggesting, and it would probably be pretty garbage, for example:
if i had a context like ['I','like'] and the model gave two predictions as samples, ['toys', 'candy'], and you add them, you see its noisy and weird. now do this max_new_tokens times, and it will probably suck.

**Comment by VatsaDev** [comment]
Yep its pure garbage:

User: Hello?
Bot: IHi, Hi, there!, I Hi, I Hi'm I, I i i I Hi I HiHi,!< I HiHi., Hi I HiHi., Hi hi hi I HiHi.HiHi.Hi ,Hi HiHi HiHi Hi HiHiHi,Hi<endbr>><humanbot>><endp><>
Is the there is any way chance to you could make the your website own a domain name? <<endn>ore<>
<botsystem>><endhuman>><humanend="Of courseastron! I<endbr>>"
>&
WhatI'm mean of to do say,? <endbr>>"
How would can I i make do i? <<endbr>><botend="=""
>
&lt
& 
lt;&ltgt;&ltgt;><endimg src height="=&1"";&
 &ltgt

crazy amounts of noise from going by 2's let alone other values, so it wont work

**Comment by mw66** [comment]
> We might be able to get multiple tokens at a time with `idx_next = torch.multinomial(probs, num_samples=1)` being set to 2 or 5, etc, gotta test that.
> 
This will get two alternative tokens for the same next position.


> but still, it won't be the same thing you're suggesting, and it would probably be pretty garbage, for example: if i had a context like ['I','like'] and the model gave two predictions as samples, ['toys', 'candy']

These two alternative tokens (for the same next position) are not meant to be appended sequentially, and this is not what I want.

What I want is to train and predict next two sequential tokens for ［n+1, n+2］ together.

**Comment by VatsaDev** [comment]
Yep I realize that, that's why i mentioned the second part. I don't think it's possible without some sort of architecture change, as most LLMs are based off of predicting the next token.

something similar is speculative decoding, where you make a smaller llm generate whole sentences for the prompt, and have a larger LLM rank them, which produces the output in sentence batches.

**Comment by smougel** [comment]
Hello @mw66 
I was thinking about that today...
Forcing the model to predict 2 tokens at a time...

But remember that the model generate only one token during the forward pass:
```
logits = self.lm_head(x) # (B,T,vocab_size)
```

To output 2 tokens, we need to modify our projection head a little bit...
`self.lm_head = nn.Linear(n_embd, vocab_size, bias=False)`
probably by multiplying the output size by 2... and that's a lot of parameters...

that could be x ---> predict y1 & y2
or
with x we predict y1 --> compute another attention pass and given x and y1 ---> output y2

I don't know if we can save a lot of computations...
